metrics/setup.go

```diff
@@ -0,0 +1,142 @@
+package metrics
+
+import (
+	"github.com/packethost/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	DHCPTotal *prometheus.CounterVec
+
+	CacherDuration           prometheus.ObserverVec
+	CacherCacheHits          *prometheus.CounterVec
+	CacherTotal              *prometheus.CounterVec
+	CacherRequestsInProgress *prometheus.GaugeVec
+
+	DiscoverDuration    prometheus.ObserverVec
+	HardwareDiscovers   *prometheus.CounterVec
+	DiscoversInProgress *prometheus.GaugeVec
+
+	JobDuration    prometheus.ObserverVec
+	JobsTotal      *prometheus.CounterVec
+	JobsInProgress *prometheus.GaugeVec
+)
+
+func Init(_ log.Logger) {
+	DHCPTotal = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "dhcp_total",
+		Help: "Number of DHCP Requests received.",
+	}, []string{"op", "giaddr"})
+
+	labelValues := []prometheus.Labels{
+		{"op": "DHCPACK", "giaddr": "0.0.0.0"},
+		{"op": "DHCPDECLINE", "giaddr": "0.0.0.0"},
+		{"op": "DHCPDISCOVER", "giaddr": "0.0.0.0"},
+		{"op": "DHCPINFORM", "giaddr": "0.0.0.0"},
+		{"op": "DHCPNAK", "giaddr": "0.0.0.0"},
+		{"op": "DHCPOFFER", "giaddr": "0.0.0.0"},
+		{"op": "DHCPRELEASE", "giaddr": "0.0.0.0"},
+		{"op": "DHCPREQUEST", "giaddr": "0.0.0.0"},
+		{"op": "reply", "giaddr": "0.0.0.0"},
```

Comment:
reply? what does this mean/how is it used?

Reply:
ok looks like this is meant for "successful" replies. But that should probably be handled via the normal dhcp response right? We should never see a DHCPOFFER coming in since we config dhcp4-go to ignore w/e we don't care about anyway.

I'm back to thinking we need to have total and responded/error separately.

Reply:
I thought the same, however I thought it would be useful to not update DHCPOFFER in case a DHCPOFFER made it to the server. Seeing that number go up could help determine that there are multiple dhcp servers on the network.

I can add a dhcp_offers generic count separately. But since this one tracks by giadder, it might be good to keep this here.

Reply:
ah just noticed dhcp_totals is also broken out by giadder, so i could do the same as dhcp_totals, and remove reply from here

Reply:
maybe add a direction then?

Reply:
switched to making op = send/recv and moved the type to a type tag.